src/modules/history/streaming-services/viaplay/ViaplayApi.ts

```diff
@@ -0,0 +1,149 @@
+
+class _ViaplayApi implements Api {
+  HOST_URL: string;
+  HISTORY_API_URL: string;
+  HISTORY_API_NEXT_PAGE_URL: string;
+  AUTH_URL: string;
+  isActivated: boolean;
+
+
+  constructor() {
+    this.HOST_URL = 'https://content.viaplay.no';
+    this.HISTORY_API_URL = `${this.HOST_URL}/pcdash-no/watched`;
+    this.AUTH_URL = `https://login.viaplay.no/api/persistentLogin/v1?deviceKey=pcdash-no`;
+    this.HISTORY_API_NEXT_PAGE_URL = this.HISTORY_API_URL;
+
+    this.isActivated = false;
+
+    this.loadHistory = this.loadHistory.bind(this);
+    this.parseHistoryItem = this.parseHistoryItem.bind(this);
+    this.loadTraktHistory = this.loadTraktHistory.bind(this);
+    this.loadTraktItemHistory = this.loadTraktItemHistory.bind(this);
+  }
+
+  async activate() {
+    await Requests.send({
+      url: this.AUTH_URL,
+      method: 'GET',
+    });
+    this.isActivated = true;
+  }
+
+  async loadHistory(nextPage: number, nextVisualPage: number, itemsToLoad: number) {
+    try {
+      if (!this.isActivated) {
+        await this.activate();
+      }
+      let isLastPage = false;
+      let items: Item[] = [];
+      const historyItems: ViaplayProduct[] = [];
+
+      do {
+        const responseText = await Requests.send({
+          url: this.HISTORY_API_NEXT_PAGE_URL,
+          method: 'GET',
+        });
+        let historyPage: ViaplayHistoryPage;
+        if (this.HISTORY_API_NEXT_PAGE_URL === this.HISTORY_API_URL) { //First initial load/page
+          const responseJson: ViaplayWatchedTopResponse = JSON.parse(responseText);
+          historyPage = responseJson._embedded['viaplay:blocks'][0];
+        } else {
+          historyPage = JSON.parse(responseText);
+        }
+        const viaplayProducts: ViaplayProduct[] = historyPage._embedded['viaplay:products'];
+
+        if (viaplayProducts && viaplayProducts.length > 0) {
+          itemsToLoad -= viaplayProducts.length;
+          historyItems.push(...viaplayProducts);
+        } else {
+          isLastPage = true;
+        }
+        nextPage += 1;
+        const url = historyPage._links?.next?.href;
+        this.HISTORY_API_NEXT_PAGE_URL = url;
+        if (!url) {
+          isLastPage = true;
+        }
+      } while (!isLastPage && itemsToLoad > 0);
+      if (historyItems.length > 0) {
+        items = historyItems.map(this.parseHistoryItem);
+      }
+      nextVisualPage += 1;
+      ViaplayStore.update({ isLastPage, nextPage, nextVisualPage, items })
+        .then(this.loadTraktHistory);
+    } catch (err) {
+      Errors.error('Failed to load Viaplay history.', err);
+      await EventDispatcher.dispatch(Events.STREAMING_SERVICE_HISTORY_LOAD_ERROR, { error: err });
+    }
+  }
+
+  parseHistoryItem(historyItem: ViaplayProduct): Item {
+    let item: Item = null;
+    const year = historyItem.content.production.year;
+    const percentageWatched = historyItem.user.progress?.elapsedPercent || 0;
+    const watchedAt = moment(historyItem.user.progress?.updated);
```

Comment:
I found one item that had no progress-field in my own history. Not quite sure what do with such items. Right now this will transform into today's date. Another approach is to remove the item entirely.